client/my-sites/checkout/upsell-nudge/purchase-modal/index.tsx

```diff
@@ -202,3 +209,28 @@ export default function PurchaseModalWrapper( props: PurchaseModalProps ) {
 		</CheckoutProvider>
 	);
 }
+
```

Comment:
`NB:`

I'm wondering about a few things:

- Are we creating a HOC because we intend to share this logic with other components?
- Is there a place further upstream where we it might be better to set the selected site? I suppose, intuitively, I wouldn't expect a leaf component like the purchase modal to be handling that operation

Reply:
That's a good question!

> Are we creating a HOC because we intend to share this logic with other components?

This component can be reusable, but as of now I am not aware of any other use cases. I'll try to explain why a new component was required.

Before this change, the code in the `<PurchaseModalWrapper/>` component looked like this:
```ts
const selectedSite = useSelector( getSelectedSite );
const siteId = useSelector( ( state ) => getSiteId( state, siteSlug ) );

// Set the selected site if it is not set already.
// This is necessary for the cart and post-purchase actions to function correctly.
const hasSelectedSiteId = siteId === selectedSite?.ID;
useEffect( () => {
	if ( ! hasSelectedSiteId && siteId ) {
		reduxDispatch( setSelectedSiteId( siteId ) );
	}
}, [ hasSelectedSiteId, reduxDispatch, siteId ] );

const cartKey = useCartKey();
const { replaceProductsInCart, ... } = useShoppingCart( cartKey);
```

Zooming in on this part:
```ts
const cartKey = useCartKey();
const { replaceProductsInCart, ... } = useShoppingCart( cartKey);
```

The `cartKey` would change whenever the [selected site id changed](https://github.com/Automattic/wp-calypso/blob/5866ce5706b220b8eeb3c8391460348813d19057/client/my-sites/checkout/use-cart-key.ts#L8), but the reference to the `replaceProductsInCart` function does not change when the `cartKey` changes. This causes cart actions to happen on an invalid cart. I am not sure if it's intentional or a bug in `useShoppingCart()` though. 🤔 (@sirbrillig can you please help in clarifying this?)

The `<EnsureSelectedSite />` component ensures that the `<PurchaseModalWrapper/>` is rendered with the corrected `selectedSiteId` and thus the correct `cartKey`.

I tried switching to `cartKey = getCartKey( { selectedSite: <site> } )` in this component to avoid setting a `selectedSiteId`, but it doesn't help as `useCartKey` is still used downstream in post-purchase hooks like `useCreatePaymentCompleteCallback`, `useRedirectOnTransactionSuccess`, etc. 

> Is there a place further upstream where we it might be better to set the selected site? I suppose, intuitively, I wouldn't expect a leaf component like the purchase modal to be handling that operation

This modal was previously used In logged-in Calypso contexts where the `selectedSiteId` is already set. In other contexts (signup/stepper), if we pass this responsibility to upstream components, it would be tough to enforce this restriction. Maybe we can throw an error if the `selectedSiteId` is not set? WDYT?